### PR TITLE
feat: add generic nvme support

### PR DESCRIPTION
**Pull request by jippi**
## What?

This is my local work, trying to solve same issue as #1233 

This PR makes NVMe transparent to rexray, as it translates between the NVMe logical device and the virtual device that is the udev symlink with the device name that rexray request EC2 to mount the EBS volume as

## Tested on 

- [x] c4
- [x] c5
- [x] c5d
- [x] m4
- [x] m5
- [x] m5d
- [x] r4
- [x] r5
- [x] r5d
- [x] i3.2xlarge

## Requirements

Require a `udev` rule to alias the NVMe device to the path that rexray expects as mount point.

A similar udev rule is build into the Amazon Linux AMI already, and trivial to add to other linux distributions (see below)

An alternative is to manage the symlinks from the rexray driver (its just a symlink), but seem like feature creep to me, considering the low effort required to make the `udev` setup work

### Example udev rule

```
# /etc/udev/rules.d/999-aws-ebs-nvme.rules
# ebs nvme devices
KERNEL=="nvme[0-9]*n[0-9]*", ENV{DEVTYPE}=="disk", ATTRS{model}=="Amazon Elastic Block Store", PROGRAM="/usr/local/bin/ebs-nvme-mapping /dev/%k", SYMLINK+="%c"
```

### Helper script for udev rules

```sh
#!/bin/bash
#/usr/local/bin/ebs-nvme-mapping
vol=$(/usr/sbin/nvme id-ctrl --raw-binary "$1" | cut -c3073-3104 | tr -s ' ' | sed 's/ $//g')
vol=${vol#/dev/}
if [[ -n "$vol" ]]; then
  echo ${vol/xvd/sd} ${vol/sd/xvd}
fi
```

Signed-off-by: Christian Winther <jippignu@gmail.com>

### CODE REVIEW

**Comment by clintkitson** [comment]
cc @justinvenus @crapworks @codenrhoden 

I do like the approach more. What does everyone feel about this path?

**Comment by jippi** [comment]
@clintkitson Can we please get the ball rolling on this? been a issue for many months now, slightly frustrating that its moving so slow :)

**Comment by clintkitson** [comment]
Thanks for the patience @jippi. Glad to see the community work through updates on this. Can you take a stab at updating the documentation that describe how to make this work in this pr?

**Comment by clintkitson** [comment]
@justinvenus Thoughts on this implementation versus the one you submitted?

**Comment by JustinVenus** [comment]
TL;DR looks good, please shipit/merge --- 

@clintkitson I haven't tested this PR, but this PR looks very straightforward and easy to maintain.  I personally would not hard code the path to the nvme-cli, but I'd worry about that nitpick if it ever became a real issue.

**Comment by clintkitson** [comment]
Just waiting on some doc updates and we will ship this in a minor release.

**Comment by jamesdh** [comment]
Assuming this get's approved and merged in the coming few days, about how long of a wait time would there be for a rexray release containing it?

**Comment by clintkitson** [comment]
We can turn this around within days, but am still waiting on some doc updates that indicate how to configure the new functionality.

**Comment by jippi** [comment]
The OP have the two steps. It's literally just putting the two files on disk and reboot / restart udev :)

**Comment by akutz** [comment]
Hi @jippi,

Once upon a time I wrote a [blog](https://blog.thecodeteam.com/2017/09/13/reproducing-rex-ray/) that outlined how to reproduce REX-Ray locally down to the binaries' MD5 hash. In other words, even without this merged it shouldn't be too difficult for y'all to use it in your environment.

**Comment by matthewmrichter** [comment]
Bump.

**Comment by jwitko** [comment]
Would really love to see this merged.

**Comment by akutz** [comment]
Hi @jippi / @matthewmrichter / @jwitko,

I'm going to merge this, but I'm holding one or all of you accountable (aka asking you nicely with a cherry on top) to please go file a PR that takes the instructions in this PR description and incorporates them into the RR docs all nice and pretty like. Once that's done, we can cut a patch release so this is available GA. At least merged you'll be able to pull it officially as the latest unstable bits.

File: .docker/plugins/Dockerfile [hunk 1]
```diff
@@ -1,7 +1,7 @@
 FROM alpine:3.6
 
 RUN apk update
-RUN apk add xfsprogs e2fsprogs ca-certificates
+RUN apk add xfsprogs e2fsprogs ca-certificates nvme-cli
```
Comment (akutz) [inline .docker/plugins/Dockerfile]: Hi @jippi, 

Please make a specific Dockerfile for AWS and add nvm-cli there, not to the generic Dockerfile.

File: libstorage/drivers/storage/ebs/executor/ebs_executor.go [hunk 2]
```diff
@@ -165,11 +166,34 @@ func (d *driver) LocalDevices(
...
+
+		// NVMe support
+		if strings.Contains(devName, "nvme") {
+			// find the EBS device name that we *think* we will mount the device as (nvme ignore this)
+			if out, err := exec.Command("/usr/sbin/nvme", "id-ctrl", "--raw-binary", devPath).Output(); err == nil {
+				// read the binary output slice and trim it
+				dev := strings.TrimSpace(string(out[3072:3104]))
+				// if the result contains a /dev/ then we got a match
+				if strings.Contains(dev, "/dev/") {
+					log.Debugf("Found symlink for '%s' -> '%s'", devName, dev)
```
Comment (akutz) [inline libstorage/drivers/storage/ebs/executor/ebs_executor.go]: Please use the context logger, not `log` directly. Also please use the structured logging as is used everywhere else in RR.

File: libstorage/drivers/storage/ebs/executor/ebs_executor.go [hunk 3]
```diff
@@ -165,11 +166,34 @@ func (d *driver) LocalDevices(
...
+					if _, err := os.Stat(dev); !os.IsNotExist(err) {
+						devName = strings.TrimLeft(dev, "/dev/")
+						devPath = dev
+					}
+				}
+			}
+		}
+
 		if !ns.DeviceRE.MatchString(devName) {
+			log.Warnf("Device '%s' do not match '%s'", devName, ns.DeviceRE)
```
Comment (akutz) [inline libstorage/drivers/storage/ebs/executor/ebs_executor.go]: See notes above about logging.

File: libstorage/drivers/storage/ebs/executor/ebs_executor.go [hunk 4]
```diff
@@ -165,11 +166,34 @@ func (d *driver) LocalDevices(
...
 		}
 		devName := fields[3]
+
+		devPath := path.Join("/dev/", devName)
+		devAlias := devPath
+
+		// NVMe support
+		if strings.Contains(devName, "nvme") {
+			// find the EBS device name that we *think* we will mount the device as (nvme ignore this)
+			if out, err := exec.Command("/usr/sbin/nvme", "id-ctrl", "--raw-binary", devPath).Output(); err == nil {
```
Comment (akutz) [inline libstorage/drivers/storage/ebs/executor/ebs_executor.go]: Please make this command path configurable with this as the default.

File: libstorage/drivers/storage/ebs/executor/ebs_executor.go [hunk 5]
```diff
@@ -165,11 +166,34 @@ func (d *driver) LocalDevices(
...
+				// read the binary output slice and trim it
+				dev := strings.TrimSpace(string(out[3072:3104]))
+				// if the result contains a /dev/ then we got a match
+				if strings.Contains(dev, "/dev/") {
+					log.Debugf("Found symlink for '%s' -> '%s'", devName, dev)
+					// if the alias / udev path exist, its a match
+					if _, err := os.Stat(dev); !os.IsNotExist(err) {
```
Comment (akutz) [inline libstorage/drivers/storage/ebs/executor/ebs_executor.go]: Please be careful when doing `os.Stat`. Use `gotils` which has the edge case handled.

**Comment by akutz** [comment]
Hi @jippi,

If you can't attend to the review comments, I'll take over the PR (keeping your commit) and get this merged by the end of the week.

**Comment by akutz** [comment]
Hi @jippi,

I just pushed [`e26cf35`](https://github.com/akutz/rexray/commit/e26cf35d00a5f237d2384b88329225b4dbc5183c) to address my review comments. Please feel free to use it to update your PR.

**Comment by jippi** [comment]
@akutz feel free to highjack it - I'm on vacation this week - thank you  :)

**Comment by akutz** [comment]
Hi @jippi / @matthewmrichter / @jwitko,

This has been merged. Please see 407253f50421326b3f9d6aa93858af9ce65ba51e for the addendums I made to the PR. They include:

* Documentation
* Configurable path to `nvme`
* More robust file detection
* Contextual logging

cc @clintkitson @codenrhoden

**Comment by jippi** [comment]
Any plans for cutting a release? :)

**Comment by jippi** [comment]
Given its taking full 9 months from initial PR to any release, and 3 months from merge and still no release, I would probably start considering REXray for deprecated and unmaintained. I'm running my own fork internally at @seatgeek now, which is this branch. Got tired of waiting around

**Comment by matthewmrichter** [comment]
Yes, we are switching to the [Cloudstor](
https://docs.docker.com/docker-for-aws/persistent-data-volumes/#shared-cloudstor-volumes) driver which is quite simple to set up and supports EFS as well. I believe it is maintained by Docker corporation themselves. Unfortunately they are dragging their heels on NVME a bit as well but hopefully that gets fixed soon.

**Comment by clintkitson** [comment]
Hello folks, we have been waiting on a community member to run a test of the unstable release. If that happens and we can get a report here we will proceed with a release.

**Comment by karlatkinson** [comment]
A few people have tested this out here https://github.com/rexray/rexray/issues/1104
Are you still looking for more feedback before releasing as stable?

**Comment by akutz** [comment]
Hi @karlatkinson / @matthewmrichter / @jippi / @ccakes,

My thanks to the community for testing this PR. Release [v0.11.4](https://github.com/rexray/rexray/releases/tag/v0.11.4) is [building](https://travis-ci.org/rexray/rexray/builds/479962434) now. I'll ping you again when the binaries are available. Thank you.

**Comment by akutz** [comment]
Hi @karlatkinson / @matthewmrichter / @jippi / @ccakes,

The complete set of the v0.11.4 binaries are now [available](https://dl.bintray.com/rexray/rexray/stable/0.11.4/). Please note that Bintray appears to be suffering from gremlins, and I've had to spend one-two hours re-running builds until the BT server will accept uploads. This means not all of the Docker images are not yet available, but they are uploading fine and the complete set should be available in the next hour.
